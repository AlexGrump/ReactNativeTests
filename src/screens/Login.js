import {Text, SafeAreaView, View} from 'react-native';
import React, {useState} from 'react';
import appStyles from '../styles/app.styles';
import Button from '../components/Button';
import Input from '../components/Input';
import FormErrors from '../components/FormErrors';

const useLoginFormState = props => {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [submited, setSubmited] = useState(false);
  const [inProgress, setInProgress] = useState(false);

  let usernameErrors = [];
  if (username !== 'user@example.com') {
    usernameErrors.push('Invalid username');
  }

  let passwordErrors = [];
  if (password !== 'password') {
    passwordErrors.push('Invalid password');
  }

  return {
    fields: {
      username: {
        value: username,
        set: setUsername,
        errors: submited ? usernameErrors : [],
      },
      password: {
        value: password,
        set: setPassword,
        errors: submited ? passwordErrors : [],
      },
    },
    inProgress: inProgress,
    submited: submited,
    submit: () => {
      setSubmited(true);
      if (!inProgress && !usernameErrors.length && !passwordErrors.length) {
        setInProgress(true);
        fetch('https://jsonplaceholder.typicode.com/users', {
          method: 'POST',
          body: JSON.stringify({
            username,
            password,
          }),
        })
          .then(response => response.json())
          .then(() => {
            setInProgress(false);
            props.navigation.replace('Home');
          })
          .catch(error => {
            setInProgress(false);
          });
      }
    },
  };
};

const Login = props => {
  const {fields, inProgress, submit} = useLoginFormState(props);

  return (
    <SafeAreaView style={appStyles.root}>
      <View style={appStyles.floatContainer}>
        <Text style={appStyles.h1}>Login</Text>
        <Input
          testID="Login.usernameInput"
          label="Username"
          value={fields.username.value}
          onChangeText={fields.username.set}
          placeholder="user@example.com"
          error={fields.username.errors.length}
          disabled={inProgress}
        />
        <Input
          testID="Login.passwordInput"
          label="Password"
          value={fields.password.value}
          onChangeText={fields.password.set}
          placeholder="password"
          secureTextEntry
          error={fields.password.errors.length}
          disabled={inProgress}
        />
        <FormErrors
          errors={[...fields.username.errors, ...fields.password.errors]}
        />
        <Button
          disabled={inProgress}
          onPress={submit}
          testID="Login.submitButton">
          Login
        </Button>
      </View>
    </SafeAreaView>
  );
};

export default Login;
