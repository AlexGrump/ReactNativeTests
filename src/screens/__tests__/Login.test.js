import React from 'react';
import {act, fireEvent, render} from '@testing-library/react-native';

import Login from '../Login';

const flushMicrotasksQueue = () => {
  return new Promise(resolve => setImmediate(resolve));
};

it('renders default elements', () => {
  const result = render(<Login />);

  expect(result.getAllByText('Login').length).toBe(2);
  result.getByPlaceholderText('user@example.com');
  result.getByPlaceholderText('password');
});

it('show form errors', () => {
  const result = render(<Login />);

  fireEvent.press(result.getByTestId('Login.submitButton'));

  result.getByText('Invalid username');
  result.getByText('Invalid password');
});

it('show invalid user name error message', () => {
  const result = render(<Login />);

  fireEvent.changeText(result.getByTestId('Login.passwordInput'), 'password');

  fireEvent.press(result.getByTestId('Login.submitButton'));

  result.getByText('Invalid username');
  expect(result.queryAllByText('Invalid password').length).toBe(0);
});

it('show invalid password error message', () => {
  const result = render(<Login />);

  fireEvent.changeText(
    result.getByTestId('Login.usernameInput'),
    'user@example.com',
  );

  fireEvent.press(result.getByTestId('Login.submitButton'));

  result.getByText('Invalid password');
  expect(result.queryAllByText('Invalid username').length).toBe(0);
});

it('handles submit with valid inputs', async () => {
  fetch.mockResponseOnce(JSON.stringify({success: true}));

  const replaceMock = jest.fn();

  const result = render(<Login navigation={{replace: replaceMock}} />);

  fireEvent.changeText(
    result.getByTestId('Login.usernameInput'),
    'user@example.com',
  );
  fireEvent.changeText(result.getByTestId('Login.passwordInput'), 'password');
  fireEvent.press(result.getByTestId('Login.submitButton'));

  await act(flushMicrotasksQueue);

  expect(fetch.mock.calls).toMatchSnapshot();
  expect(replaceMock).toBeCalledWith('Home');
});
