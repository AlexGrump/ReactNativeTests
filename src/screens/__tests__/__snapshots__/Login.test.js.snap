// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles submit with valid inputs 1`] = `
Array [
  Array [
    "https://jsonplaceholder.typicode.com/users",
    Object {
      "body": "{\\"username\\":\\"user@example.com\\",\\"password\\":\\"password\\"}",
      "method": "POST",
    },
  ],
]
`;
